src/hypothesis/extra/pandas/impl.py

```diff
@@ -0,0 +1,361 @@
+            If an explicit index is provided then max_size may be at most the
+            length of the index. If an index strategy is provided then whenever
+            the drawn index is too short max_size will merely be reduced.
+
+        index: If not None, an index for the resulting DataFrame. If provided,
+            the min_size and max_size must be compatible with the index size
+            but will otherwise be ignored.
+
+    """
+
+    index = build_index(
+        draw, index, min_size, max_size
+    )
+
+    if rows_or_columns is not None:
+        if rows is not None:
+            if columns is not None:
+                raise InvalidArgument(
+                    'At most two of rows, columns, and rows_or_columns can be '
+                    'provided.'
+                )
+            else:
+                columns = rows_or_columns
+        elif columns is not None or isinstance(
+            rows_or_columns, st.SearchStrategy
+        ):
+            rows = rows_or_columns
+        else:
+            columns = rows_or_columns
+
+    if isinstance(columns, column):
+        columns = (columns,)
+
+    if columns is None:
+        if rows is None:
+            raise InvalidArgument(
+                'At least one of rows and columns must be provided'
+            )
+
+        if len(index) > 0:
+            return pandas.DataFrame(
+                [draw(rows) for _ in index],
+                index=index
+            )
+        else:
+            # If we haven't drawn any rows we need to draw one row and then
+            # discard it so that we get a consistent shape for the the
+            # DataFrame.
+            base = draw(st.shared(
+                rows.map(lambda x: pandas.DataFrame([x])),
+                key=('hypothesis.extra.pandas.row_shape', rows),
+            ))
+            return base.drop(0)
+
+    column_names = []
+    dtypes = []
+    strategies = []
+    categorical_columns = []
+
+    for i, c in enumerate(columns):
+        st.check_type(column, c, 'columns[%d]' % (i,))
+        name = c.name
+        if name is None:
+            name = i
+        dtype = c.dtype
+        elements = c.elements
+        if dtype is None:
+            dtype = draw(dtype_for_elements_strategy(elements))
+        if is_category_dtype(dtype):
+            dtype = np.dtype(object)
+            categorical_columns.append(name)
+        else:
+            dtype = np.dtype(dtype)
+
+        if elements is None:
+            elements = npst.from_dtype(dtype)
+
+        column_names.append(name)
+        strategies.append(elements)
+        dtypes.append(dtype)
+
+    if len(set(column_names)) < len(column_names):
+        counts = Counter(column_names)
+        raise InvalidArgument(
+            'columns definition contains duplicate column names: %r' % (
+                sorted(c for c, n in counts.items() if n > 1)))
+
+    zeroes = [0] * len(index)
```

Comment:
It would be faster to use https://docs.scipy.org/doc/numpy/reference/generated/numpy.zeros.html here.